### PR TITLE
[WIP] Add scscore implemented using PyTorch

### PR DESCRIPTION
**Pull request by BalajiAI**
## Description

I've implemented scscore architecture using PyTorch as referenced in https://github.com/deepchem/deepchem/issues/2863. I created the architecture using PyTorch nn module. I need to implement the TorchModel part (I've no idea on how to do it, need some more time to figure out).

Also, the original keras implementation seems outdated. I didn't implement scscore based on keras implementation, but based on the original paper. So I'm just making this PR to get feedback.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/scscore.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import torch
+import torch.nn as nn
+from deepchem.data import NumpyDataset
+from deepchem.feat import CircularFingerprint
+from deepchem.models.torch_models.torch_model import TorchModel
+from deepchem.models.losses import L1Loss
+
+class ScScore(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:8]: DeepChem follows 2-spacing at present. Make sure to format the code properly

File: deepchem/models/torch_models/scscore.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        for hidden_layer in self.hidden_layers:
+            x = hidden_layer(x)
+            if self.dropout > 0:
+                x = nn.Dropout(p=self.dropout)(x)
+            x = nn.ReLU()(x)
+        output = self.output_layer(x)
+        output = nn.Sigmoid()(output)
+        return output
+        
+class ScScoreModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:41]: Can you add a docstring here?

File: deepchem/models/torch_models/scscore.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    
+    def __init__(self, n_features=1024, layer_sizes=[300,300,300,300,300], dropout = 0, **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:43]: Can you add type annotations to the arguments here?

File: deepchem/models/torch_models/scscore.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import torch
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:1]: Could you add unit tests for these classes?

**Review by rbharath — COMMENTED** [review]
Good start! We will need unit tests and documentation before this code will be ready to merge. I suggest coming by one of the DeepChem office hours to discuss with us

File: deepchem/models/torch_models/scscore.py (hunk 2 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py:43]: Sure. But here, [normalizing flows](https://github.com/deepchem/deepchem/blob/master/deepchem/models/torch_models/normalizing_flows_pytorch.py) the person just implemented normalizing_flows using torch.nn, I think he needs to add normalizing_flows_model implemented using torch_model. Am I right? Or just implementing models using torch.nn is enough?

**Comment by BalajiAI** [comment]
> Good start! We will need unit tests and documentation before this code will be ready to merge. I suggest coming by one of the DeepChem office hours to discuss with us
thank you for the helpful feedback.

Also, does the link mentioned here - https://forum.deepchem.io/t/announcing-the-deepchem-office-hours/293 is valid still? Or do you use different meeting ids?

File: deepchem/models/torch_models/scscore.py (hunk 3 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py:48]: Added for ScScore as of now

File: deepchem/models/torch_models/scscore.py (hunk 4 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py:1]: Done !

**Comment by BalajiAI** [comment]
@arunppsg can you review?

File: deepchem/models/tests/test_scscore_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  import torch.nn as nn
+  from deepchem.models.torch_models.scscore import ScScore
+  has_torch = True
+except:
+  has_torch = False
+  
+  
+@unittest.skipIf(not has_torch, 'torch is not installed')
+@pytest.mark.torch
+def test_scscore_pytorch():
```
Comment (arunppsg) [inline deepchem/models/tests/test_scscore_pytorch.py:17]: A test for model should cover model overfit and model reload.

File: deepchem/models/torch_models/scscore.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import torch.nn as nn
+from deepchem.data import NumpyDataset
+from deepchem.feat import CircularFingerprint
+from deepchem.models.torch_models.torch_model import TorchModel
+from deepchem.models.losses import L1Loss
+
+
+class ScScore(nn.Module):
```
Comment (arunppsg) [inline deepchem/models/torch_models/scscore.py:9]: A suggestion: use `SCScore` . From [pep-0008](https://peps.python.org/pep-0008/),

> when using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

File: deepchem/models/torch_models/scscore.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    ScScore paper - https://pubs.acs.org/doi/abs/10.1021/acs.jcim.7b00622
```
Comment (arunppsg) [inline deepchem/models/torch_models/scscore.py:11]: SCScore is more than a feed-forward network. Refer paper for the writing documentation.

File: deepchem/models/torch_models/scscore.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            if self.dropout > 0:
+                x = nn.Dropout(p=self.dropout)(x)
+            x = nn.ReLU()(x)
+        output = self.output_layer(x)
+        output = nn.Sigmoid()(output)
+        return output
+        
+        
+class ScScoreModel(TorchModel):
+    """
```
Comment (arunppsg) [inline deepchem/models/torch_models/scscore.py:44]: Add docstrings and include model in `docs/source/api_reference/models.rst`

File: deepchem/models/torch_models/scscore.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    
+    """
+    
+    def __init__(self, n_features=1024, layer_sizes=[300,300,300,300,300], dropout = 0, **kwargs):
+        self.n_features = n_features
+        model = ScScore(n_features, layer_sizes, dropout)
+        output_types = ['prediction']
+        super(ScScoreModel, self).__init__(
+             model, L1Loss(), output_types=output_types, **kwargs)
```
Comment (arunppsg) [inline deepchem/models/torch_models/scscore.py:53]: Original implementation uses HingeLoss

File: deepchem/models/torch_models/scscore.py (hunk 6 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py]: Ok done.

File: deepchem/models/torch_models/scscore.py (hunk 7 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py]: Sure. I'll recheck.

File: deepchem/models/tests/test_scscore_pytorch.py (hunk 5 above)
Comment (BalajiAI) [reply deepchem/models/tests/test_scscore_pytorch.py:17]: Oh!

File: deepchem/models/torch_models/scscore.py (hunk 8 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py:44]: I'll add at the end.

File: deepchem/models/torch_models/scscore.py (hunk 9 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py:53]: Oops

File: deepchem/models/torch_models/scscore.py (hunk 7 above)
Comment (BalajiAI) [reply deepchem/models/torch_models/scscore.py]: I read the paper. It is a FFNN.

From the paper -
The model, implemented in Python 2.7
using Tensorflow, 30 is a feed forward neural network consisting of five hidden layers with 300
hidden nodes each, ReLU activation with bias, and a single sigmoid output.

File: deepchem/models/tests/test_scscore_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    import torch
+    from deepchem.models.torch_models.scscore import SCScoreModel
+    has_torch = True
+except:
+    has_torch = False
+
+
+@pytest.mark.torch
+def test_scscore_reload():
+    """Tests the model reload capability"""
```
Comment (rbharath) [inline deepchem/models/tests/test_scscore_pytorch.py:17]: Need to run yapf

File: deepchem/models/torch_models/scscore.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+class SCScore(nn.Module):
+    """
+    SCScore is a Feed-Forward Neural network with 5 layers.
+    Each layer consists of only 300 neurons.
+    """
+    
+    def __init__(self, n_features: int, layer_sizes: list, dropout: int, **kwargs) -> None:
+        """
+        Arguments
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:17]: The docstring formatting should be in numpy style
